### PR TITLE
QR code detection

### PR DESCRIPTION
**Pull request by pablospe**
This adds QR code detection in Capture sessions.

```bash
> python -m scantools.run_qrcode_detection --help
usage: run_qrcode_detection.py [-h] --capture_path CAPTURE_PATH [--sessions [SESSIONS ...]] [--navvis_path NAVVIS_PATH] [--visualization | --no-visualization]
                               [--txt_format | --no-txt_format] [--json_format | --no-json_format]

This script is used for QR code detection. The script saves the QR map (list of all QR codes) in the `CAPTURE_PATH/proc/qrcodes/` directory. The QR map can be saved in
both TXT and JSON formats. The TXT format is used for compatibility with the rest of CAPTURE format. The JSON format is used for convince but disabled by default. The
script also saves the detected QR codes for each image. These are saved in the `CAPTURE_PATH/proc/qrcodes/` directory with the same name as the image file, but with the
`.qrcodes.txt` suffix. These files are used to load the QR codes if they have already been detected for the image. This is useful when we want to re-run the script
without having to detect the QR codes again.

optional arguments:
  -h, --help            show this help message and exit
  --capture_path CAPTURE_PATH
                        Path to the capture. If it doesn't exist it will process with tile format `none` and export the capture to this path.
  --sessions [SESSIONS ...]
                        [Optional] List of sessions to process. If not provided, all sessions in the `capture_path` are processed. Useful when processing only specific
                        sessions.
  --navvis_path NAVVIS_PATH
                        [Optional] Specifies NavVis data path. This argument is ignored if `--capture_path` exists, which is useful when the data has already been
                        converted to the CAPTURE format and the original NavVis data is no longer available.
  --visualization, --no-visualization
                        Write out MeshLab visualization. Default: True. Pass --no-visualization to set to False. (default: True)
  --txt_format, --no-txt_format
                        Write out QR maps in txt format. Default: True. (default: True)
  --json_format, --no-json_format
                        Write out QR maps in json format. Default: False. (default: False)
```

### CODE REVIEW

File: scantools/proc/rendering.py [hunk 1]
```diff
@@ -18,12 +18,18 @@
...
-    p2d = np.mgrid[:h, :w].reshape(2, -1)[::-1].T.astype(np.float32)
+    # If p2d (2D points) is not provided, we compute rays for the entire image.
+    if p2d is None:
+        w, h = np.array((camera.width, camera.height)) // stride
+        p2d = np.mgrid[:h, :w].reshape(2, -1)[::-1].T.astype(np.float32)
+        num_rays = h*w
+    else:
+        num_rays = len(p2d)
+
+    origins = np.tile(center.astype(np.float32)[None], (num_rays, 1))
```
Comment (sarlinpe) [inline scantools/proc/rendering.py:32]:
```suggestion
    origins = np.tile(center.astype(np.float32)[None], (len(p2d), 1))
```

File: scantools/qr/detector.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
+import csv
```
Comment (sarlinpe) [inline scantools/qr/detector.py:1]: This folder should maybe go into `proc/` ?

File: scantools/proc/rendering.py (hunk 1 above)
Comment (pablospe) [reply scantools/proc/rendering.py]: done. Thanks!

File: scantools/qr/detector.py (hunk 2 above)
Comment (pablospe) [reply scantools/qr/detector.py]: Good, I moved to `proc/qrcode`.

**Review by sarlinpe — APPROVED** [review]

File: scantools/proc/qrcode/detector.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            if header != self.__csv_header():
+                raise ValueError(
+                    "The CSV header file does not match the expected format."
+                )
+
+            self.qrcodes = []
+            for row in reader:
+                qr_code = {
+                    "id": row[0],
+                    "points2D": [
```
Comment (mihaidusmanu) [inline scantools/proc/qrcode/detector.py:117]: `list(map(float, row[1 : 9]))`

File: scantools/run_qrcode_detection.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        # closely. As a result, we don't need tiling, which could potentially
+        # split the QR code across multiple tiles.
+        tiles_format = "none"
+        run_navvis_to_capture.run(navvis_path, capture, tiles_format)
+
+    # If `sessions` is not provided, run for all sessions in the `capture_path`.
+    if sessions is None:
+        sessions = capture.sessions.keys()
+
+    mesh_id = "mesh"
```
Comment (mihaidusmanu) [inline scantools/run_qrcode_detection.py:57]: Maybe use the simplified since raybender will require a lot of RAM for the full one?

**Review by mihaidusmanu — APPROVED** [review]

File: scantools/proc/qrcode/detector.py (hunk 3 above)
Comment (pablospe) [reply scantools/proc/qrcode/detector.py:117]: Thanks! It is not exactly the same. With an example, the current code gives you:
```
[[2477.0, 3341.0], [2492.0, 3542.0], [2686.0, 3554.0], [2667.0, 3350.0]]
```
and your suggestion:
```
[2477.0, 3341.0, 2492.0, 3542.0, 2686.0, 3554.0, 2667.0, 3350.0]
```

**Pull request merged**